src/spade/transformer/ContextualTemporalTraversal.java

```diff
@@ -0,0 +1,272 @@
+                    AbstractVertex edgeChild = edge.getChildVertex();
+                    AbstractVertex edgeParent = edge.getParentVertex();
+                    final String edgeHash = edge.bigHashCode();
+                    edgeMap.get(edgeChild.bigHashCode()).get("parentEdges").put(edgeHash, edge);
+                    edgeMap.get(edgeParent.bigHashCode()).get("childEdges").put(edgeHash, edge);
+                }
+                return edgeMap;
+            }
+
+        @Override
+        public Graph transform(Graph graph, ExecutionContext context) {
+                Set<AbstractVertex> currentLevel = new HashSet<>();
+                /*
+                * Pick a start vertex to begin traversal. Pass it to transform method in SPADE Query client.
+                * Example: $1 = $2.transform(ContextualTemporalTraversal, "order=timestamp", $startVertex, 'descendant')
+                */
+                List<AbstractVertex> startGraphVertexSet = new ArrayList<AbstractVertex>(context.getSourceGraph().vertexSet());
+                if (startGraphVertexSet.isEmpty()) {
+                        logger.log(Level.SEVERE, "Source graph is empty. Cannot initiate traversal.");
+                        return null;
+                }
+                AbstractVertex startVertex = startGraphVertexSet.get(0);// Select the first vertex from the list as the starting point for traversal
+                HashMap<String, HashMap<String, HashMap<String, AbstractEdge>>> edgeMap = setEdgeMap(graph);
+                Integer levelCount = 0;
+
+                currentLevel.add(startVertex);
+                Graph finalGraph = new Graph();
+                logger.log(Level.INFO, "Traversal started. Initiating the process with the selected start vertex.");
+                while (!currentLevel.isEmpty()) {
+                        logger.log(Level.INFO, "Processing vertices at level: " + levelCount + ". Checking child vertices for traversal.");
+                        Graph adjacentGraph = null;
+                        HashMap<String, HashMap<String, HashMap<String, AbstractEdge>>> finalEdgeMap = setEdgeMap(finalGraph);
+                        for (AbstractVertex node : currentLevel) {
+                                // get children of current level nodes
+                                // timestamp of subsequent edges > than current
+                                adjacentGraph = getAllChildrenForAVertex(node, adjacentGraph, finalGraph, graph, levelCount, edgeMap, finalEdgeMap);
+                        }
+                        logger.log(Level.INFO, "Children vertices received for current level " + levelCount + ". Moving to process them.");
+                        if (!adjacentGraph.vertexSet().isEmpty()) {// If children exists
+                                logger.log(Level.INFO, "Adding child vertices to graph and preparing for DFS on the next level.");
+                                // Add children in graph and run DFS on children
```

Comment:
Minor: Comments says DFS instead of BFS in some places.

Reply:
Thanks for pointing that out. I have corrected the comments. Note: The top level comment in the file still says "DFS", since it refers to the DFS mentioned in the original WATSON paper.

Reply:
Thank you.